core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -395,31 +382,100 @@ private void enterIdleMode() {
     channelStateManager.gotoState(IDLE);
   }
 
-  // Must be run from channelExecutor
-  private void cancelIdleTimer() {
-    if (idleModeTimerFuture != null) {
-      idleModeTimerFuture.cancel(false);
-      idleModeTimer.cancelled = true;
-      idleModeTimerFuture = null;
-      idleModeTimer = null;
+  static final class Rescheduler {
+    private final Runnable runnable;
+    private final ChannelExecutor exec;
+    private final ScheduledExecutorService scheduler;
+
+    private ScheduledFuture<?> wakeUp;
+
+    // state
+    private long runAt;
+    private boolean enabled;
+
+    Rescheduler(Runnable r, ChannelExecutor exec, ScheduledExecutorService scheduler) {
+      this.runnable = r;
+      this.exec = exec;
+      this.scheduler = scheduler;
+    }
+
+    // must be called from channel executor
+    void reschedule(long delay, TimeUnit timeUnit) {
+      long delayNanos = timeUnit.toNanos(delay);
+      long newRunAt = ticker.read() + delayNanos;
+      if (newRunAt - runAt < 0 || !enabled) {
+        enabled = true;
+        if (wakeUp != null) {
+          wakeUp.cancel(false);
+        }
+        wakeUp = scheduler.schedule(new FutureRunnable(), delayNanos, TimeUnit.NANOSECONDS);
+      }
+      runAt = newRunAt;
+    }
+
+    // must be called from channel executor
+    void cancel(boolean permanent) {
+      enabled = false;
+      if (permanent && wakeUp != null) {
+        wakeUp.cancel(false);
+        wakeUp = null;
+      }
+    }
+
+    private final class FutureRunnable implements Runnable {
+
+      @Override
+      public void run() {
+        exec.executeLater(new ChannelFutureRunnable());
+        exec.drain();
+      }
+    }
+
+    private final class ChannelFutureRunnable implements Runnable {
+
+      @Override
+      public void run() {
+        if (!enabled) {
+          return;
+        }
+        long now = ticker.read();
+        if (runAt - now > 0) {
+          wakeUp = scheduler.schedule(new FutureRunnable(), runAt - now,  TimeUnit.NANOSECONDS);
+        } else {
+          enabled = false;
+          wakeUp = null;
+          runnable.run();
+        }
+      }
+    }
+
+    /** Time source representing nanoseconds since fixed but arbitrary point in time. */
+    interface Ticker {
```

Comment:
FYI: Guava's ticker is an abstract class, and marked `@Beta`.  I can't use it here, so I define an interface.

Reply:
I toyed with this some using Stopwatch. I think it ends up fine (no unnecessary now() calls in the normal paths), and it is a stable API.

Reply:
Switched to using stopwatch.